File: csharp/DesiArchitect.DesignPatterns/06_Adapter/After/BluedartAdapter.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Adapter.After;

public class BluedartAdapter : ILogisticsPartner
{
    private readonly BluedartClient _client;

    public BluedartAdapter(BluedartClient client) => _client = client;

    public TrackingId Ship(Parcel parcel)
    {
        Console.WriteLine($"  🔄 BluedartAdapter: translating Parcel -> BluedartConsignmentRequest");
        // The adapter owns the translation - our domain Parcel never knows the vendor's shape.
        var request = new BluedartConsignmentRequest(parcel.OrderId, parcel.Address, parcel.WeightKg);
        var awb = _client.CreateConsignment(request);
        return new TrackingId(awb.Number);
    }
}

File: typescript/src/06-adapter/after/bluedart-adapter.ts
import type { Parcel, TrackingId } from '../../shared/models.js';
import type { ILogisticsPartner } from './ilogistics-partner.js';
import { BluedartClient } from './bluedart-client.js';

export class BluedartAdapter implements ILogisticsPartner {
  constructor(private readonly client: BluedartClient) {}

  ship(parcel: Parcel): TrackingId {
    console.log('  🔄 BluedartAdapter: translating Parcel -> BluedartConsignmentRequest');
    const request = {
      refId: parcel.orderId,
      destination: parcel.address,
      weight: parcel.weightKg,
    };
    const awb = this.client.createConsignment(request);
    return { value: awb.number };
  }
}